lake/branch.go

```diff
@@ -90,252 +54,223 @@ func (b *Branch) Load(ctx context.Context, r zio.Reader, date nano.Ts, author, m
-			b, err = b.pool.OpenBranchByID(ctx, b.Parent)
-			if err != nil {
-				return nil, 0, err
-			}
-		default:
-			return nil, 0, err
+func commonAncestor(a, b []ksuid.KSUID) ksuid.KSUID {
+	m := make(map[ksuid.KSUID]struct{})
+	for _, id := range a {
+		m[id] = struct{}{}
+	}
+	for _, id := range b {
+		if _, ok := m[id]; ok {
+			return id
 		}
 	}
+	return ksuid.Nil
 }
 
-func (b *Branch) openParent(ctx context.Context) (*Branch, error) {
-	parent, err := b.pool.OpenBranchByID(ctx, b.Parent)
-	if err != nil {
-		return nil, fmt.Errorf("branch %s/%s could not access parent branch %s: %w", b.pool.Name, b.Name, b.Parent, err)
-	}
-	return parent, nil
-}
+type constructor func(parent *branches.Config, retries int) (*commits.Object, error)
 
-func (b *Branch) snapParentAt(ctx context.Context, at journal.ID) (commit.View, error) {
-	parent, err := b.openParent(ctx)
-	if err != nil {
-		return nil, err
-	}
-	snap, err := parent.snapshotAt(ctx, at)
-	if err != nil {
-		return nil, fmt.Errorf("branch %s/%s could not access parent branch %s@%d: %w", b.pool.Name, b.Name, b.Parent, at, err)
+func (b *Branch) commit(ctx context.Context, create constructor) (ksuid.KSUID, error) {
+	// A commit must append new state to the tip of the branch while simultaneously
+	// upating the branch pointer in a trasactionally consistent fashion.
+	// For example, if we compute a commit object based on a certain tip commit,
+	// then commit that object after another writer commits in between,
+	// the commit object may be inconsistent against the intervening commit.
+	//
+	// We do this update optimistically and ensure this consistency with
+	// a loop that builds the commit object based on the presumed parent,
+	// then moves the branch pointer to the new commit but, using a constraint,
+	// only succeeds when the presumed parent is atomically consistent
+	// with the branch update.  If the contraint, fails will loop a number
+	// of times till it succeeds, or we give up.
+	for retries := 0; retries < maxCommitRetries; retries++ {
+		config, err := b.pool.branches.LookupByName(ctx, b.Name)
+		if err != nil {
+			return ksuid.Nil, err
+		}
+		object, err := create(config, retries)
+		if err != nil {
+			return ksuid.Nil, err
+		}
+		if err := b.pool.commits.Put(ctx, object); err != nil {
+			return ksuid.Nil, fmt.Errorf("branch %q failed to write commit object: %w", b.Name, err)
+		}
+		// Set the branch pointer to point to this commit object
+		// and stash the current commit (that will become the parent)
+		// in a local for the constraint check closure.
+		parent := config.Commit
+		config.Commit = object.Commit
+		parentCheck := func(e journal.Entry) bool {
+			if entry, ok := e.(*branches.Config); ok {
+				return entry.Commit == parent
+			}
+			return false
+		}
+		if err := b.pool.branches.Update(ctx, config, parentCheck); err != nil {
+			if err == journal.ErrConstraint {
+				// Parent check failed so try again.
+				continue
+			}
+			return ksuid.Nil, err
+		}
```

Comment:
Should we delete the object we wrote to the commit store if we fail / need to retry, or will this get cleaned up with garbage collection?

Reply:
Very good point!